File: src/main/java/me/basiqueevangelist/datapackify/JsonUtils.java
package me.basiqueevangelist.datapackify;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.recipe.ShapedRecipe;
import net.minecraft.util.Identifier;
import net.minecraft.util.JsonHelper;
import net.minecraft.util.registry.Registry;

public final class JsonUtils {
    private JsonUtils() {

    }

    public static JsonElement get(JsonObject obj, String name) {
        if (!obj.has(name))
            throw new IllegalArgumentException("Missing " + name + "in JSON object!");

        return obj.get(name);
    }

    public static <T> T getRegistryItem(Registry<T> reg, String id) {
        return reg.getOrEmpty(new Identifier(id)).orElseThrow(() ->
            new IllegalArgumentException("'" + id + "' is not a member of registry " + reg.getKey().getValue())
        );
    }

    public static ItemStack makeItemStack(JsonElement el) {
        if (el.isJsonPrimitive()) {
            return new ItemStack(JsonHelper.asItem(el, "<item stack>"));
        } else {
            return ShapedRecipe.outputFromJson(JsonHelper.asObject(el, "<item stack>"));
        }
    }
}

File: src/main/kotlin/me/basiqueevangelist/datapackify/JsonUtils.kt
package me.basiqueevangelist.datapackify

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.mojang.brigadier.StringReader
import com.mojang.brigadier.exceptions.CommandSyntaxException
import net.minecraft.item.ItemStack
import net.minecraft.nbt.StringNbtReader
import net.minecraft.util.Identifier
import net.minecraft.util.JsonHelper
import net.minecraft.util.registry.Registry

object JsonUtils {
    fun get(obj: JsonObject, name: String): JsonElement {
        return obj.get(name) ?: throw IllegalArgumentException("Missing $name")
    }

    fun <T> getRegistryItem(reg: Registry<T>, id: String): T {
        return reg.getOrEmpty(Identifier(id)).orElseThrow {
            IllegalArgumentException("'$id' is not a member of registry ${reg.key.value}")
        }
    }

    fun makeItemStack(from: JsonElement): ItemStack {
        return if (from.isJsonPrimitive) ItemStack(JsonHelper.asItem(from, "<item stack>")) else {
            val obj = JsonHelper.asObject(from, "<item stack>")
            val ist = ItemStack(
                JsonHelper.getItem(obj, "item"),
                JsonHelper.getInt(obj, "count", 1)
            )
            if (obj.has("nbt")) {
                val tag = JsonHelper.getString(obj, "nbt")
                try {
                    ist.setTag(StringNbtReader.parse(tag))
                } catch (cse: CommandSyntaxException) {
                    throw java.lang.IllegalArgumentException("Could not parse NBT tag: $cse")
                }
            }
            ist
        }
    }
}